### ISSUE TITLE
strange Pool::collect behaviour

### ISSUE THREAD

**Issue by MacJudge** [krakjoe/pthreads#491]
Hi @krakjoe.

I'm new to multi-threading in PHP and started to try and modify some examples with the release of pthreads 3.0 (now using v3.0.5) and PHP 7.0 (now using RC3) at my x64 Windows PC using PHP CLI.

Until now I have some problems understanding how to use the Pool-class and how to get the results of my workers collected, when they are finished. Are there any good examples with complex (nested) objects? When I build my objects and try to evaluate them after Pool::shutdown, my objects are destroyed and I get a corresponding error message. But before Pool::shutdown they are not yet built.

So I assumed, that Pool::collect might be my friend and looked at your example at http://stackoverflow.com/a/32187103
Unfortunately, I discovered that there isn't always one message "Collecting ..." for each thread, making this useless. I don't know, whether this is intended, but it might be a bug.

Playing around with Pool::collect and your example I came up with the following code:
```php
<?php
// poolsize doesn't matter
$pool = new Pool(4);
// just submit one task to pool
$pool->submit(new class extends Collectable {
  public function __construct() {
    // simulate some complex object to be created
    for($c = 0; $c < 10000000; $c++);
  }

  public function run() {
    // simple work
    echo "Hello World\r\n";
    $this->setGarbage();
  }
});
// check what we can collect from pool
for($c = 0; $c < 5; $c++) {
  $p = $pool->collect(function(Collectable $work) use (&$garbage){return $garbage = $work->isGarbage();});
  echo $c.': '.$p.' '.($garbage ? 'Y' : 'N')."\r\n";
}
// shutdown the pool
$pool->shutdown();
?>
```
The expected output would be:
> Hello World
> 0: 1 N
> 1: 0 Y
> 2: 0 Y
> 3: 0 Y
> 4: 0 Y

But sometimes I get something like:
> 0: 0 N
> 1: 1 N
> 2: 1 N
> 3: 1 N
> Hello World
> 4: 1 N

The important thing is the return of "0" from Pool::collect at first position, which breaks the proposed
```php
while($pool->collect(function(Collectable $work){return $work->isGarbage();}))
  continue;
```
As this is your demo-code, it seems like an unintended bug.

If you have problems reproducing this behaviour, perhaps trying some slow computer might help (I use some old Core 2 Duo).

**Comment by krakjoe** [comment]
The purpose of ```Pool::collect``` is only to allow garbage collection to occur, it's nothing to do with reaping results. Workers have to maintain a reference to objects on their stack while they are being executed, collect gives you a way to cleanup the references previously executed. When you shutdown a pool, all workers are destroyed, along with their stacks.

When you submit() execution is scheduled in an available worker, but its not guaranteed to occur instantly, it depends how busy the workers are, so the assumption that after submit() you should be able to read the Collectables data is wrong.

Have a look at the following code:

	<?php
	class Mine extends Collectable {
		public function run() {
			$this->synchronized(function(){
				$this->data = rand(1,10);
				$this->notify(); # someone might be waiting for this data, time to wake them up
			});
			$this->setGarbage();
		}
	}

	$mine = new Mine();

	$pool = new Pool(4);
	$pool->submit($mine);

	/* do other work */

	$mine->synchronized(function() use($mine) {
		while (!$mine->data) # if no data available
			$mine->wait();    # wait for it

		var_dump($mine->data); # when execution gets to here
							   # we know we have data
	});

	/* done with $mine */
	unset($mine);

	/* allow garbage collection to occur */
	$pool->collect(function($task){
		return $task->isGarbage();
	});

	/* destroy the pool */
	$pool->shutdown();
	?>

That should clear up the confusion.

**Comment by MacJudge** [comment]
Thanks for your help and the example code. The confusion is cleared (at least a bit), but I still have problems to access my objects.

I modified your code to work with some nested object and ended in a race condition with garbage collection (I'm not sure about this, but this is, what I suppose):
```php
<?php
class MyArray extends Threaded {}
class MyContainer extends Threaded {
  public $myarray;
  public function __construct(MyArray $myarray) {
    $this->myarray = $myarray;
  }
  public function add(int $i) {
    $this->myarray[] = $i;
  }
}
class Mine extends Collectable {
  public function run() {
    $this->synchronized(function(){
      // build complex object
      $mycontainer = new MyContainer(new MyArray());
      for($c = 0; $c < 10; $c++)
        $mycontainer->add(mt_rand(1, 10));
      // assign object to output
      $this->data = $mycontainer;
      $this->notify(); # someone might be waiting for this data, time to wake them up
    });
    $this->setGarbage();
  }
}

$mine = new Mine();

$pool = new Pool(4);
$pool->submit($mine);

/* do other work */
// when PC is too fast, just add another '0'
// to be sure, that garbage collection wins and the following code fails
for($c = 0; $c < 10000000; $c++);

$mine->synchronized(function() use($mine) {
    while (!$mine->data) # if no data available
        $mine->wait();    # wait for it

    var_dump($mine->data); # when execution gets to here
                           # we know we have data
});

/* done with $mine */
unset($mine);

/* allow garbage collection to occur */
$pool->collect(function($task){
    return $task->isGarbage();
});

/* destroy the pool */
$pool->shutdown();
?>
```
When garbage collection wins, I get the following exception:
```
PHP Fatal error:  Uncaught RuntimeException: pthreads detected an attempt to connect
to an object which has already been destroyed in Test.php:38
Stack trace:
#0 [internal function]: {closure}()
#1 Test.php(43): Threaded->synchronized(Object(Closure))
#2 {main}
  thrown in Test.php on line 38
```

Do you have another hint, how to get access to my objects?

**Comment by krakjoe** [comment]
> Do you have another hint, how to get access to my objects?

Try to remember there is more than one context: They aren't actually your objects, they belong to the child thread that created them (where new was invoked), when it finishes, they are destroyed, hence you raise an exception when you try to access them in the main context.

I'm not sure why the example code is so complicated, there is no need to nest objects like that.

The closest code to your that will work as you expect is this:

	<?php
	class MyArray extends Threaded {}

	class MyContainer extends Threaded {
	  public $myarray;
	  public function __construct(MyArray $myarray) {
		$this->myarray = $myarray;
	  }
	  public function add(int $i) {
		$this->myarray[] = $i;
	  }
	}

	class Mine extends Collectable {

	  public function __construct() {
		# The constructor is called in the Thread that invoked "new"
	  	$this->container = 
			new MyContainer(new MyArray());
	  }

	  public function run() {
		$this->synchronized(function(){
		  // build complex object
		  for($c = 0; $c < 10; $c++)
		    $this->container->add(mt_rand(1, 10));
		  $this->data = true;
		  $this->notify(); # someone might be waiting for this data, time to wake them up
		});
		$this->setGarbage();
	  }
	}

	$mine = new Mine();

	$pool = new Pool(4);
	$pool->submit($mine);

	/* do other work */
	// when PC is too fast, just add another '0'
	// to be sure, that garbage collection wins and the following code fails
	for($c = 0; $c < 10000000; $c++);

	$mine->synchronized(function() use($mine) {
		while (!$mine->data) # if no data available
		    $mine->wait();    # wait for it

		var_dump($mine->container); # when execution gets to here
		                       # we know we have data
	});

	/* done with $mine */
	unset($mine);

	/* allow garbage collection to occur */
	$pool->collect(function($task){
		return $task->isGarbage();
	});

	/* destroy the pool */
	$pool->shutdown();
	?>

But nicer, would be this:

	<?php
	class Container extends Collectable {

	  public function __construct() {
	  	$this->array = new Threaded();
	  }

	  public function run() {
		$this->synchronized(function(){
		  // build complex object
		  for($c = 0; $c < 10; $c++)
		    $this->array[] = mt_rand(1, 10);

		  $this->done = true;
		  $this->notify(); # someone might be waiting for this data, time to wake them up
		});
		$this->setGarbage();
	  }

	  public $done;
      public $array;
	}

	$mine = new Container();

	$pool = new Pool(4);
	$pool->submit($mine);

	/* do other work */
	// when PC is too fast, just add another '0'
	// to be sure, that garbage collection wins and the following code fails
	for($c = 0; $c < 10000000; $c++);

	$mine->synchronized(function() use($mine) {
		while (!$mine->done) # if no data available
		    $mine->wait();    # wait for it

		var_dump($mine->array); # when execution gets to here
		                       # we know we have data
	});

	/* done with $mine */
	unset($mine);

	/* allow garbage collection to occur */
	$pool->collect(function($task){
		return $task->isGarbage();
	});

	/* destroy the pool */
	$pool->shutdown();
	?>

Even nicer would be like this:

	<?php
	class Container extends Collectable {

	  public function __construct(Threaded $array) {
	  	$this->array = $array;
	  }

	  public function run() {
		$this->synchronized(function(){
		  // build complex object
		  for($c = 0; $c < 10; $c++)
		    $this->array[] = mt_rand(1, 10);

		  $this->done = true;
		  $this->notify(); # someone might be waiting for this data, time to wake them up
		});
		$this->setGarbage();
	  }

	  public $done;
      public $array;
	}

	$array = new Threaded();
	$mine = new Container($array);

	$pool = new Pool(4);
	$pool->submit($mine);

	/* do other work */
	// when PC is too fast, just add another '0'
	// to be sure, that garbage collection wins and the following code fails
	for($c = 0; $c < 10000000; $c++);

	$mine->synchronized(function() use($mine) {
		while (!$mine->done) # if no data available
		    $mine->wait();    # wait for it

		var_dump($mine->array); # when execution gets to here
		                       # we know we have data
	});

	/* done with $mine */
	unset($mine);

	/* allow garbage collection to occur */
	$pool->collect(function($task){
		return $task->isGarbage();
	});

	/* destroy the pool */
	$pool->shutdown();

	var_dump($array);
	?>

Always try to make it as simple as possible, and remember an object belong to whichever context constructed it.

**Comment by MacJudge** [comment]
Thanks again for your help and to keep this short, I'd like to say "Thank you very much" for your work. You're giving PHP users on Windows the possibility to use multi-threading and I was really exited, when your new API for PHP 7 was available.

Usually, I use procedural programming with only one single thread. I have an efficient algorithm, which heavily relies on recursion and where information is stored in multi-level arrays. As I thought of some easy speed-up by multi-threading, I rewrote most of my arrays to be objects and many of my functions to be methods to use your extension. With your help, I finally came up with a not-crashing version, that gets the same results like my single-threaded version.

Unfortunately, there is a giant leak in handles and corresponding memory, which makes my program useless until now, but I'll do some more research to get this fixed and hopefully end up with a really working version. And if I fail, there is always the possibility to use your API to call different instances of my single-threaded script on command line and collect and combine their results in one control-program. ;)
